controllers/clustermodule_reconciler.go

```diff
@@ -0,0 +1,255 @@
+		return reconcile.Result{}, err
+	}
+
+	var errList []error
+	clusterModuleSpecs := []infrav1.ClusterModule{}
+
+	for _, mod := range ctx.VSphereCluster.Spec.ClusterModules {
+		curr := mod.TargetObjectName
+		if mod.ControlPlane {
+			curr = appendKCPKey(curr)
+		}
+		if obj, ok := objectMap[curr]; !ok {
+			// delete the cluster module as the object is marked for deletion
+			// or already deleted.
+			if err := r.ClusterModuleService.Remove(ctx, mod.ModuleUUID); err != nil {
+				ctx.Logger.Error(err, "failed to delete cluster module for object",
+					"name", mod.TargetObjectName, "moduleUUID", mod.ModuleUUID)
+				errList = append(errList, err)
+			}
+			delete(objectMap, curr)
+		} else {
+			// verify the cluster module
+			exists, err := r.ClusterModuleService.DoesExist(ctx, obj, mod.ModuleUUID)
+			if err != nil {
+				errList = append(errList, err)
+			}
+			// append the module and object info to the VSphereCluster object
+			// and remove it from the object map since no new cluster module
+			// needs to be created.
+			if exists {
+				clusterModuleSpecs = append(clusterModuleSpecs, infrav1.ClusterModule{
+					ControlPlane:     obj.IsControlPlane(),
+					TargetObjectName: obj.GetName(),
+					ModuleUUID:       mod.ModuleUUID,
+				})
+				delete(objectMap, curr)
+			} else {
+				ctx.Logger.Info("module for object not found",
+					"moduleUUID", mod.ModuleUUID,
+					"object", mod.TargetObjectName)
+			}
+		}
+	}
+	if len(errList) > 0 {
+		ctx.Logger.Error(kerrors.NewAggregate(errList), "errors reconciling cluster modules for cluster",
+			"namespace", ctx.VSphereCluster.Namespace, "name", ctx.VSphereCluster.Name)
+	}
+
+	errList = []error{}
+	for _, obj := range objectMap {
+		moduleUUID, err := r.ClusterModuleService.Create(ctx, obj)
+		if err != nil {
+			ctx.Logger.Error(err, "failed to create cluster module for target object", "name", obj.GetName())
+			errList = append(errList, err)
+		}
+		clusterModuleSpecs = append(clusterModuleSpecs, infrav1.ClusterModule{
+			ControlPlane:     obj.IsControlPlane(),
+			TargetObjectName: obj.GetName(),
+			ModuleUUID:       moduleUUID,
+		})
+	}
+	ctx.VSphereCluster.Spec.ClusterModules = clusterModuleSpecs
+
+	if len(errList) > 0 {
+		err = kerrors.NewAggregate(errList)
+		ctx.Logger.Error(err, "errors reconciling cluster modules for cluster",
+			"namespace", ctx.VSphereCluster.Namespace, "name", ctx.VSphereCluster.Name)
+	}
+
+	switch {
+	case err != nil:
+		conditions.MarkFalse(ctx.VSphereCluster, infrav1.ClusterModulesAvailableCondition, infrav1.ClusterModuleSetupFailedReason, clusterv1.ConditionSeverityWarning, err.Error())
+	case err == nil && len(clusterModuleSpecs) > 0:
+		conditions.MarkTrue(ctx.VSphereCluster, infrav1.ClusterModulesAvailableCondition)
+	default:
+		conditions.Delete(ctx.VSphereCluster, infrav1.ClusterModulesAvailableCondition)
+	}
+	return reconcile.Result{}, err
+}
+
+func (r Reconciler) ToAffinityInput(obj client.Object) []reconcile.Request {
```

Comment:
this is only used in this file so it should likely be `toAffinityInput`

Reply:
Doen